### PR TITLE
Implement site-specific config file loading.

### LINKED ISSUE #3968: Multisite installs do not load `drush.yml`
**Describe the bug**
`drush.yml` is not loaded for multisite installs.

**To Reproduce**
Put a `drush.yml` in `web/sites/default/drush.yml`

**Expected behavior**
The path to that drush.yml file should appear in `drush status`

**Actual behavior**
Nothing at all.

**Workaround**
Unknown; maybe set `uri` in settings.php. Drush.yml might be used for other things though.

### System Configuration
| Q               | A
| --------------- | ---
| Drush version?  | 9.6.0-rc3

**Additional information**
Drush is [attempting to add the configuration in ConfigLocator](https://github.com/drush-ops/drush/blob/9.6.0-rc3/src/Config/ConfigLocator.php#L282
), but something is amis.

Fixing this was previously attempted and rejected in #3278.

### LINKED ISSUE #4324: drush.yml isn't discovered in web/sites/default/drush.yml
According to [example.drush.yml](https://github.com/drush-ops/drush/blob/master/examples/example.drush.yml) it should be possible to add a `drush.yml` file these locations:
```
# Directories and Discovery
#
# Rename this file to drush.yml and copy it to one of the places listed below
# in order of precedence:
#
# 1.  Drupal site folder (e.g. sites/{default|example.com}/drush.yml).
# 2.  Drupal /drush and sites/all/drush folders, or the /drush folder
#       in the directory above the Drupal root.
# 3.  In any location, as specified by the --config (-c) option.
# 4.  User's .drush folder (i.e. ~/.drush/drush.yml).
# 5.  System wide configuration folder (e.g. /etc/drush/drush.yml or C:\ProgramData\Drush\drush.yml).
```
I would prefer to use the first option, by simply adding a file to an existing folder, but it doesn't seem to work:
- `web/sites/default/drush.yml`

These work:
- `drush/drush.yml`
- `web/sites/all/drush/drush.yml`

**To Reproduce**
Add a `web/sites/default/drush.yml` file, which isn't discovered.

### System Configuration
| Q               | A
| --------------- | ---
| Drush version  | 9.7.2
| Drupal version | 8.8.2
| PHP version     | 7.2.24
| OS             | Ubuntu with Lando

### LINKED ISSUE #4551: sites/default/drush.yml no longer loaded?
The documentation seems to suggest sites/default/drush.yml is a location where the base url can be set (for auto-generated links like login links).  Is this no longer a valid config location?  I had hoped to be able to set the base url for all of our developer's lando boxes to be the same, ergo:
options:
  uri: 'https://kstat.lndo.site/'

but the location of the other config files don't seem like they're really appropriate for version control.  Has this changed, is sites/default/drush.yml no longer valid?

### System Configuration
| Q               | A
| --------------- | ---
| Drush version?  | 9.72
| Drupal version? | 8.86
| PHP version     | 7.1
| OS?             | Linux

### LINKED ISSUE #4638: Configure URI for default site
**Background/motivation**
Follow on from #4551.  Each Drupal site can respond to many URIs but effectively has a "canonical URI" used in sitemaps, metadata, system mails etc.  It's essential for drush to set the URI correctly especially for running cron.  Cron jobs may send mails, create search indices etc and they need to be done with the right base URI.  For all of the secondary sites, the canonical URI is implicitly defined from the directory name sites/XXX.  For the default site we need another way to set it.  The file sites/all/drush/drush.yml affects all sites so it's no good (unless you have just one site).

**Describe the solution you'd like**
The obvious answer to me is to allow sites/default/drush.yml to set options:uri.   I agree that sites/default/drush.yml is too late to affect which site gets bootstrapped.  However it seems entirely realistic and very convenient to use it to define the URI that gets passed to Drupal for the default site.

**Describe alternatives you've considered**
In 4551 @weitzman suggested to use environment variables, especially when developers are using different hostname.  However it seems like a nightmare for a team of admins trying to manage dozens of small sites .  The canonical site URL of a site should be defined in a single central location not per developer.  When we want to run an operation on a group of sites, the environment would somehow have to be swapped in between each one.  Drush does an excellent job of solving exactly this problem for every other parameter about a site, so it seems a shame if it can't help with

### PR DESCRIPTION
**Pull request by richardbporter**
Resolves #3968 

Resolves #4324
Resolves #4638
Resolves #4551

### CODE REVIEW

**Comment by weitzman** [comment]
The active site is not known during Preflight. That happens during the SITE phase of Drush bootstrap (where we call Drupal's API to determine it).

**Comment by weitzman** [comment]
Drupal needs to consult sites.php and the incoming Request URL in order to determine active site.

**Comment by richardbporter** [comment]
> The active site is not known during Preflight. That happens during the SITE phase of Drush bootstrap (where we call Drupal's API to determine it).

Can you point me to where that is happening?

**Comment by richardbporter** [comment]
I went with Preflight because @greg-1-anderson indicated in https://github.com/drush-ops/drush/issues/3968#issuecomment-469335072 that this could be fixed "per what we're doing in #3966." Perhaps I misunderstood that.

It doesn't look like Application or any of the bootstrap classes have access to ConfigLocator. I'm assuming I would need to inject it somewhere.

**Comment by greg-1-anderson** [comment]
> Drupal needs to consult sites.php and the incoming Request URL in order to determine active site.

What @weitzman means by this is that $uri is going to be something like `https://example.com`, whereas what we need during preflight is the name of the directory that contains the settings.php file. There are various rules that Drupal employs to figure out which configuration file to use. Historically, Drush has always relied on Drupal's logic to find the settings.php file. The advantage of doing this is that we are not at risk of selecting a different directory than Drupal does. Also, the implementation is not trivial (sites.php is consulted, there are port and subdirectory selectors, etc).

It is not an accident that ConfigLocator is not available to Application or the bootstrap classes. Config loading should happen in preflight, by design. If we allowed config loading to happen later, it could introduce inconsistencies. This isn't an impossible situation, e.g. php has config settings that can only be affected in php.ini, not from ini_set, & etc., but for Drush, it would be better to do everything in preflight.

Clearly one of these two constraints must change in order to support configuration files for multisite setups. I'm ambivalent about which way is better. The comments in the code are set up to assume that the directory mapping code is going to be duplicated, but adding more configuration in bootstrap might actually be preferable. I'd try to avoid using the ConfigLocator class for this, though, if possible.

**Comment by richardbporter** [comment]
Thanks for the info. I never pass the scheme when using the `uri` option but that does make sense. I'm still not sure how best to proceed. I'll try to get back to this when I can.

**Comment by richardbporter** [comment]
Could Preflight do something similar to what the cache rebuild command is doing to access `DrupalKernel::findSitePath`?

https://github.com/drush-ops/drush/blob/8307a4e7dae9c63a279d8b59b9dca6c1973576c2/src/Commands/core/CacheCommands.php#L210-L217

**Comment by weitzman** [comment]
It could, but that would make the current phased bootstrap useless. We phase in our bootstrap with Preflight (comes even before the phases), ROOT, SITE, DB, and FULL phases. Your suggestion is to do SITE during Preflight. This is an OK suggestion, but it needs thought and would need to go in Drush 11.

**Comment by weitzman** [comment]
What config do you want to read from a site specific file?

**Comment by richardbporter** [comment]
`drush.paths.cache-directory` - somewhat related to https://github.com/drush-ops/drush/issues/3592, https://github.com/drush-ops/drush/issues/3968#issuecomment-517323144 and https://github.com/acquia/blt/issues/4009.

It would be ideal if each multisite could load its own `drush.yml` file and set that path. Setting the environment variable per request isn't feasible for every scenario, e.g. Drush alias, uri option, AC cloud hook, etc.

**Comment by weitzman** [comment]
You can set env variables in drush site aliases - an old feature that came back this fall in #4173. So if you run your command with `drush @multisite` you can set DRUSH_PATHS_CACHE_DIRECTORY just for that @multisite. See https://github.com/drush-ops/drush/blob/master/examples/example.site.yml#L171-L189

**Comment by richardbporter** [comment]
> So if you run your command with drush @multisite you can set DRUSH_PATHS_CACHE_DIRECTORY just for that @MultiSite

BLT runs Drush commands using `@self` + the `--uri` option. I know the problem I'm trying to solve here is not a Drush problem but this would offer more flexibility and a potential workaround.

This works but I'm not seeing any configuration set in a multisite `drush.yml` file reported in `drush status`, ex. `drush -l mysite.com status --field drush-cache-directory`.

**Comment by richardbporter** [comment]
I can work to add test coverage for this if you think this approach is ok.

**Comment by weitzman** [comment]
I dont think its OK, given greg's comment here.

**Comment by richardbporter** [comment]
I thought that comment indicated bootstrap might be preferable:
> but adding more configuration in bootstrap might actually be preferable. I'd try to avoid using the ConfigLocator class for this, though, if possible.

Is the issue with ConfigLocator or just doing it during bootstrap?

**Comment by weitzman** [comment]
I'm going to defer to @greg-1-anderson for that question.

**Comment by greg-1-anderson** [comment]
I think that doing it during the bootstrap is preferable; however, the bootstrap classes are not allowed to use preflight classes, so you'll have to add this without the ConfigLocator class.

Take a look at ConfigLocator::addConfigPaths() and write a similar method.

It is too late to add more include paths, so you can ignore the useMergeStrategyForKeys stuff. Of course you should be asking Drupal for the location of the `sites` directory, and find the `drush.yml` relative to that path.

It would be ok to only allow a single location, so you can ignore the "candidates" stuff. After that, you just need the part after "Complete config import", except that you won't be able to use `processedConfigPaths`. You might need to write extra code to manage the config paths so that the sites drush.yml file appears in the list of processed config files.

Handling the "recursive case" would be a nice touch, but it would be an acceptable limitation to not do it.

**Comment by richardbporter** [comment]
Sorry, just getting back around to this. Let me know what you think.

I know you said not to use the ConfigLocator class - should setting the SITE_CONTEXT be omitted?

**Comment by richardbporter** [comment]
This is working for me but I'm having trouble getting a functional test to pass. Not sure why.

**Comment by richardbporter** [comment]
I'm having other CoreTest failures locally so I commented that out for now. 
```
1) Unish\CoreCase::testOptionsUriRequestUrl
Exception: Unexpected exit code 1 (expected 0) for command:
Error Output:
================

In Drupal.php line 130:
                                                                               
  \Drupal::$container is not initialized yet. \Drupal::setContainer() must be  
   called with a real container.                                               
                                                                               

unit-eval [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-d|--debug] [-y|--yes] [--no] [--remote-host REMOTE-HOST] [--remote-user REMOTE-USER] [-r|--root ROOT] [-l|--uri URI] [--simulate] [--pipe] [-D|--define DEFINE] [--notify [NOTIFY]] [--druplicon] [--xh-link XH-LINK] [--] <command> <code>


1) Unish\CoreCase::testOptionsUri
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     'test.uri' => Array (
         0 => 'http://test.uri'
-        1 => 'sites/dev'
+        1 => 'sites/default'


2) Unish\InitCommandCase::testInitCommand
Failed asserting that '[Filesystem\FilesystemStack] mkdir ["/Users/me/.drush"]\n
```

**Comment by richardbporter** [comment]
Looking for help on this. This seems to be working with my manual testing, unless I'm overlooking something, but I'm stuck on the failing test I introduced.

**Comment by weitzman** [comment]
Code looks ok to me. Any ideas @greg-1-anderson?

**Comment by richardbporter** [comment]
Stepping through this test again. Seems like the sut/sites/dev/drush.yml file is set in the `drush status` output. Nothing seems to be loaded from it though. The `drush.paths.cache-directory` config is still set to `sandbox/home/.drush/cache` which I think is the default. I'm not seeing any other config files setting that.

One difference to other config file locations is that DrupalBoot8::bootstrapDoDrupalSite is not using a fully-qualified path. I'm not sure if that is the problem or not.

**Comment by richardbporter** [comment]
I think the issue with the test was that I was setting the cache directory to an unwritable location. It seems like `DrushConfig::cache` attempts to make the directory and silently move on to the next candidate if it can't. A log message might help there.

https://github.com/drush-ops/drush/blob/dbdb6733655231687d8ab68cdea6bf9fedbd0562/src/Config/DrushConfig.php#L77-L94

**Comment by richardbporter** [comment]
Just a ping on this one. :)

I checked "Allow edits and access to secrets by maintainers" if any maintainers want to jump in here.

**Comment by weitzman** [comment]
Code still looks good to me, and tests pass.

I changed this PR to the new 11.x base branch but that introduces conflicts so reverted. Please do that when time permits.

**Comment by richardbporter** [comment]
Thanks. I accepted all the changes from the 11.x branch for both of those files as they were not modified in this PR.

File: src/Boot/DrupalBoot8.php
Comment (richardbporter) [reply src/Boot/DrupalBoot8.php]: @greg-1-anderson Do you think this will work?

Comment (greg-1-anderson) [reply src/Boot/DrupalBoot8.php]: Yeah, it should work.

**Comment by richardbporter** [comment]
Hm. I might need some help with those CI failures. They appear to be unrelated but I could be missing something.

**Comment by greg-1-anderson** [comment]
The CircleCI tests were working 8h ago, but started failing 5h ago. Seems unrelated, but I did not try to reproduce.

**Comment by greg-1-anderson** [comment]
I'll be appveyor is failing because they upgraded their PHP to 8.0, and we're still trying to use 7.4.  Last I looked, appveyor only had one version of PHP available that would work, but you still had to specify it by version, so builds always break whenever they upgrade. I tried moving to GitHub actions a while back, but had some challenges related to the mysql database & gave up.

**Review by greg-1-anderson — APPROVED** [review]
LGTM

**Comment by weitzman** [comment]
Merged into 11.x. Don't think we should try to backport this as 10.x will soon be unsupported. Thanks for sticking with this for so long 💪

**Comment by richardbporter** [comment]
Thank you both!

**Pull request merged**
